---
title: "Preprocess data CM"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE) 
```

## Load data

```{r spliting data}
RNGkind(sample.kind = "Rounding")
set.seed(4)
setwd("./")
rm(list=ls())
require(dplyr)
getwd()
load("../../../../data/cardiomyopathy_alternative/hcm/hcm_cohort.RData")
load("../../../../data/cardiomyopathy_alternative/dcm/hvol_cm.RData")

#check whether there are intersection between hvol variants and icc mutations 
#exclude the intersections one with conflicting interpretations (hvol variants present in pathogenic data sets)
mybpc3_cohort<-subset(hcm_cohort,gene=="MYBPC3")
overlap_key<-intersect(mybpc3_cohort[which(mybpc3_cohort$pathogenic==1),]$key,hvol_cm$key)
mybpc3_cohort<-subset(mybpc3_cohort,!is.element(key,overlap_key))
hvol_cm<-subset(hvol_cm,!is.element(key,overlap_key))

mybpc3_cohort<-subset(mybpc3_cohort,select = -c(disease))

train<-rbind(mybpc3_cohort,hvol_cm)  

#only consider rare variants
train<-subset(train,(gnomAD_AF<=0.001 | is.na(gnomAD_AF)))

#exclude genes that only variants in hvol 
exclude_gene<-setdiff(unique(train[which(train$pathogenic==0),]$gene),unique(train[which(train$pathogenic==1),]$gene))
train<-subset(train,!is.element(gene,exclude_gene))

#remove duplicated rows
train<-train[!duplicated(train),]
train<-train[!duplicated(train$key),]

#remove unwanted columns and change the value type into numeric
train<-subset(train,select=-c(integrated_confidence_value,Eigen_coding_or_noncoding))
train[train=="."] <- NA

train[,11:42]<-data.matrix(train[,11:42])
train[,11:42]<-data.frame(train[,11:42])
train$SIFT_score=as.numeric(train$SIFT_score)
train$Polyphen2_HDIV_score<-as.numeric(train$Polyphen2_HDIV_score)
train$Polyphen2_HVAR_score<-as.numeric(train$Polyphen2_HVAR_score)
train$LRT_score=as.numeric(train$LRT_score)
train$MCAP=as.numeric(train$MCAP)
train$SiPhy_29way_logOdds=as.numeric(train$SiPhy_29way_logOdds)

train<-subset(train,gene=="MYBPC3")

save(train,file="../../../../data/cardiomyopathy_alternative/mybpc3_hcm/mybpc3_raw_train.RData")

load("../../../../data/cardiomyopathy_alternative/mybpc3_hcm/mybpc3_raw_train.RData")

#split the data set into 2/3 for training and 1/3 for testing and remain the class distribution 
test<-rbind(sample_n(train[which(train$pathogenic==1),],length(which(train$pathogenic==1))*1/3),
            sample_n(train[which(train$pathogenic==0),],length(which(train$pathogenic==0))*1/3))
test_id<-row.names(test)

train<-train[!rownames(train) %in% test_id,]

train_pathogenic<-subset(train,is.element(HGVSc,hcm_cohort[which(hcm_cohort$pathogenic==1),]$HGVSc))
train_benign<-subset(train,pathogenic==0)
train<-rbind(train_pathogenic,train_benign)
train<-subset(train,gene=="MYBPC3")


test_pathogenic<-subset(test,is.element(HGVSc,hcm_cohort[which(hcm_cohort$pathogenic==1),]$HGVSc))
test_benign<-subset(test,pathogenic==0)
test<-rbind(test_pathogenic,test_benign)
test<-subset(test,gene=="MYBPC3")

save(train,file="../../../../data/cardiomyopathy_alternative/mybpc3_hcm/mybpc3_train.RData")
save(test,file="../../../../data/cardiomyopathy_alternative/mybpc3_hcm/mybpc3_holdout_test.RData")

#Supplementary Table 3
table(train$pathogenic)
table(test$pathogenic)
table(rbind(train,test)$pathogenic)

#Supplementary Table 4
table(train$pathogenic,train$gene)
table(test$pathogenic,test$gene)
```


## Preprocess the data

 - Conditional Mean imputation for missing values
 - Normalization for the features in the training data set and record the corresponding mean and variation

```{r}


#######
# Imputation:
# impute_mean: mean vector used to replace the missing values from columns 14:39
# with function mean_imputation
# 
# PARASCOREzSCORE: missing value replaced as 0 and add additional dummy variable PARASCORE_exist to denote the existence of PARASCOREzSCORE
# mis_badness and MPC: missing value replaced as -1 and add addtional dummy variable mis_badness_exist to denote the existence of mis_badness
# train_imputed: raw features before normalization
# 
#######

train<-subset(train,select=-c(LRT_score))
train_imputed<-train
##For Column 14 to 39, missing is completely random. the missing values are replaced with conditional mean. 
##Mean imputation
impute_features<-colnames(train_imputed[,14:38])

impute_mean<-apply(train_imputed[,14:38],2,function(x){mean(mean(x[which(train$pathogenic==0)],na.rm=T),mean(x[which(train$pathogenic==1)],na.rm=T))})

#names of the columns which need imputation
names<-colnames(train_imputed[,14:38])
mean_imputation<-function(data,mean_vec,names){
  for(i in 1:ncol(data)){
    data[which(is.na(data[,i])==TRUE),i]<-mean_vec[names[i]]
  }
  return(data)
}

train_imputed[,14:38]<-mean_imputation(train_imputed[,14:38],impute_mean,names)

##For Column PARASCOREzSCORE, mis_badness and MPC (missing values means not applicable, replace with mean and add one more indicator variable to indicate the missingness
train_imputed$PARASCORE_exist<-ifelse(is.na(train_imputed$PARASCOREzSCORE),0,1)
train_imputed$PARASCOREzSCORE[which(is.na(train_imputed$PARASCOREzSCORE)==TRUE)]<-0

##missense badness 
train_imputed$mis_badness_exist<-ifelse(is.na(train_imputed$mis_badness),0,1)
train_imputed$mis_badness[is.na(train_imputed$mis_badness)]<--1

##MPC: use -1 to indicate NA
train_imputed$MPC[is.na(train_imputed$MPC)]<--1

#######
# Normalization:
# normalize Columns 11-39 and save the mean and std used for normalization
# with function z_normalization
#######


train_scaled<-train_imputed
scale_mean<-apply(train_scaled[,11:38],2,mean)
scale_std<-apply(train_scaled[,11:38],2,sd)

z_normalization<-function(data,mean_feature,std_feature){
  for(i in 1:ncol(data)){
    data[,i]=(data[,i]-mean_feature[i])/std_feature[i]
  }
  return(data)
}

train_scaled[,11:38]<-z_normalization(train_scaled[,11:38],scale_mean,scale_std)


train_scaled$pathogenic<-as.factor(train_scaled$pathogenic)

#remove the columns with constant values
constant_column<-apply(train_scaled,2,function(x){ifelse(length(unique(x))==1,TRUE,FALSE)})
train_scaled<-train_scaled[,-c(which(constant_column==TRUE))]

train_scaled<-sample_n(train_scaled,nrow(train_scaled))


#columns with just features:
feature_column<-colnames(train_scaled[,!(names(train_scaled) %in% c("CHROM","POS","REF","ALT","gene","AF_Adj","gnomAD_AF","HGVSc","HGVSp","key"))])

#genes assoiated with the disease:
disease_gene<-unique(train_scaled$gene)


#save output to two RData files:
#preprocess: save mean and std for feature columns and functions in order to impute and normalize new samples 
#train_scaled: training table including variant information  plus key (i.e., short form of genomic coordinate "1_XXXXXXXXXXX_A_G")
save(impute_features,impute_mean,scale_mean,scale_std,mean_imputation,z_normalization,feature_column,disease_gene,train_imputed,file="../../../../data/cardiomyopathy_alternative/mybpc3_hcm/preprocess.RData")
save(train_scaled,file="../../../../data/cardiomyopathy_alternative/mybpc3_hcm/train_scaled.RData")

```