---
title: "preprocess"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## load data

```{r cohort}
RNGkind(sample.kind = "Rounding")
rm(list=ls())
require(dplyr)
require(ggplot2)
set.seed(123)
setwd("./")
load("../../../data/arrhythmia/arm_hvol.RData")
load("../../../data/arrhythmia/arm_clinvar.RData")


#check whether there are intersection between hvol variants and icc mutations 
#exclude the intersections one with conflicting interpretations (hvol variants present in pathogenic data sets)
overlap_key<-intersect(arm_clinvar[which(arm_clinvar$pathogenic==1),]$key,arm_hvol$key)
arm_clinvar<-subset(arm_clinvar,!is.element(key,overlap_key))
arm_hvol<-subset(arm_hvol,!is.element(key,overlap_key))


#exclude the variant that has conflicting intepretation in Brugada and LQTS
#arm_clinvar<-subset(arm_clinvar,HGVSc!="ENST00000333535.4:c.5455G>A")

train<-rbind(arm_clinvar,arm_hvol) 
numeric_column<-function(data,column){
data[,column]<-as.numeric(as.character(data[,column]))
}

character_column<-c("SIFT_score","Polyphen2_HDIV_score","Polyphen2_HVAR_score","LRT_score","MutationTaster_score","MutationAssessor_score","FATHMM_score","PROVEAN_score","VEST3_score","CADD_raw","DANN_score","fathmm_MKL_coding_score","MetaSVM_score","MetaLR_score","MCAP","PARASCOREzSCORE")
train[,character_column]<-sapply(character_column,function(x){numeric_column(train,x)})

train<-subset(train,select=-c(integrated_confidence_value,Eigen_coding_or_noncoding))

table(train$pathogenic)

train<-subset(train,(gnomAD_AF<=0.001 | is.na(gnomAD_AF)))

exclude_gene<-setdiff(unique(train[which(train$pathogenic==0),]$gene),unique(train[which(train$pathogenic==1),]$gene))

train<-subset(train,!is.element(gene,exclude_gene))

train<-train[!duplicated(train$key),]

train[train=="."] <- NA

row.names(train)<-seq(1,nrow(train),1)
raw_train<-train
table(raw_train$pathogenic)

test<-rbind(sample_n(raw_train[which(raw_train$pathogenic==1),],length(which(raw_train$pathogenic==1))*1/3),sample_n(raw_train[which(raw_train$pathogenic==0),],length(which(raw_train$pathogenic==0))*1/3))
test_keys<-test$key

train<-subset(raw_train,!is.element(key,test_keys))
table(train$pathogenic,train$gene)
table(train$pathogenic)
table(test$pathogenic,test$gene)
table(test$pathogenic)

save(test,file="../../../data/arrhythmia/arm_holdout_test.RData")
save(train,file="../../../data/arrhythmia/arm_train.RData")

```



## Preprocess the data

 - Conditional Mean imputation for missing values
 - Normalization for the features in the training data set and record the corresponding mean and variation


```{r}

train_imputed<-train

impute_features<-colnames(train_imputed[,14:39])

##For Column 14 to 39, missing is completely random. the missing values are replaced with conditional mean. 
##Mean imputation

impute_mean<-apply(train_imputed[,14:39],2,function(x){mean(mean(x[which(train$pathogenic==0)],na.rm=T),mean(x[which(train$pathogenic==1)],na.rm=T))})

#The names of the columns needs imputation
names<-colnames(train_imputed[,14:39])
mean_imputation<-function(data,mean_vec,names){
  for(i in 1:ncol(data)){
    data[which(is.na(data[,i])==TRUE),i]<-mean_vec[names[i]]
  }
  return(data)
}

train_imputed[,14:39]<-mean_imputation(train_imputed[,14:39],impute_mean,names)

##For Column PARASCOREzSCORE, mis_badness and MPC (missing values means not applicable, replace with mean and add one more indicator variable to indicate the missingness
train_imputed$PARASCORE_exist<-ifelse(is.na(train_imputed$PARASCOREzSCORE),0,1)
train_imputed$PARASCOREzSCORE[which(is.na(train_imputed$PARASCOREzSCORE)==TRUE)]<-0

##missense badness [0,1]
train_imputed$mis_badness_exist<-ifelse(is.na(train_imputed$mis_badness),0,1)
train_imputed$mis_badness[is.na(train_imputed$mis_badness)]<--1
##MPC >0
train_imputed$MPC[is.na(train_imputed$MPC)]<--1

#Normalize the data 
#Recompute the mean of features after imputation

train_scaled<-train_imputed
scale_mean<-apply(train_scaled[,11:39],2,mean)
scale_std<-apply(train_scaled[,11:39],2,sd)

z_normalization<-function(data,mean_feature,std_feature){
  for(i in 1:ncol(data)){
    data[,i]=(data[,i]-mean_feature[i])/std_feature[i]
  }
  return(data)
}

train_scaled[,11:39]<-z_normalization(train_scaled[,11:39],scale_mean,scale_std)


train_scaled$pathogenic<-as.factor(train_scaled$pathogenic)

#remove the columns with constant values
constant_column<-apply(train_scaled,2,function(x){ifelse(length(unique(x))==1,TRUE,FALSE)})
train_scaled<-train_scaled[,-c(which(constant_column==TRUE))]
train_scaled<-sample_n(train_scaled,nrow(train_scaled))

#columns with just features:
feature_column<-colnames(train_scaled[,!(names(train_scaled) %in% c("CHROM","POS","REF","ALT","gene","AF_Adj","gnomAD_AF","HGVSc","HGVSp","key"))])

#genes assoiated with the disease:
disease_gene<-unique(train_scaled$gene)

#save output to two RData files:
#preprocess: save mean and std for feature columns and functions in order to impute and normalize new samples 
#train_scaled: training table including variant information  plus key (i.e., short form of genomic coordinate "1_XXXXXXXXXXX_A_G")
save(impute_features,impute_mean,scale_mean,scale_std,mean_imputation,z_normalization,feature_column,disease_gene,train_imputed,file="../../../data/arrhythmia/preprocess.RData")
save(train_scaled,file="../../../data/arrhythmia/train_scaled.RData")

```